libs/repositories/src/repos/ExpiredOrdersRepository/ExpiredOrdersRepositoryPostgres.ts

```diff
@@ -0,0 +1,69 @@
+import { Pool, QueryResult } from 'pg';
+import {
+  ExpiredOrder,
+  ExpiredOrdersContext,
+  ExpiredOrdersRepository,
+  ParsedExpiredOrder
+} from './ExpiredOrdersRepository';
+import { getOrderBookDbPool } from '../../datasources/orderBookDbPool';
+import { bytesToHexString } from '../../utils/bytesUtils';
+import { parseExpiredOrder } from './expiredOrdersUtils';
+
+export class ExpiredOrdersRepositoryPostgres implements ExpiredOrdersRepository {
+  async fetchExpiredOrdersForAccounts(context: ExpiredOrdersContext): Promise<ParsedExpiredOrder[]> {
+    const { chainId, accounts } = context;
+
+    const prodDb = getOrderBookDbPool('prod', chainId);
+    const barnDb = getOrderBookDbPool('barn', chainId);
+
+    const prodExpiredOrdersResult = await this.fetchExpiredOrdersFromDb(context, prodDb);
+    const barnExpiredOrdersResult = await this.fetchExpiredOrdersFromDb(context, barnDb);
+
+    const allExpiredOrders = [...(prodExpiredOrdersResult?.rows || []), ...(barnExpiredOrdersResult?.rows || [])];
+
+    const accountsMap = accounts.reduce<Record<string, 1 | undefined>>((acc, account) => {
+      acc[account.toLowerCase()] = 1;
+      return acc
+    }, {})
+
+    return allExpiredOrders.reduce<ParsedExpiredOrder[]>((acc, order) => {
+      if (accountsMap[bytesToHexString(order.owner).toLowerCase()]) {
+        acc.push(parseExpiredOrder(order))
+      }
+
+      return acc
+    }, [])
+  }
+
+  private async fetchExpiredOrdersFromDb(context: ExpiredOrdersContext, db: Pool): Promise<QueryResult<ExpiredOrder> | null> {
+    const { accounts, lastCheckTimestamp, nowTimestamp } = context;
+
+    if (accounts.length === 0) return null;
+
+    const query = `
+      WITH trade_sums AS (
+        SELECT
```

Comment:
this do a table scan of the whole trades. Try to filter earlier using a strong filter that is an index. Then you can join by orderId to the trades and do this grouping

Reply:
I optimized the query, thanks!
https://github.com/cowprotocol/bff/pull/155/commits/62970391b5e9fd6014db1728d6e3ba434e608bd1